[MLIR][Linalg] Add pass to convert linalg.generic back to named ops

javedabsar1 opened the pull request:
Add a new pass `--linalg-specialize-generic-ops` which lifts, where possible, linalg.generic to named ops.
Much like `-linalg-generalize-named-ops` lowers named ops to linalg.generic .
Also add patterns to recognize contractions which can be specialized from linalg.generic to named op:  `linalg.{batch_}?matmul{_transpose_(a|b)}?`
Patterns to recognize elementwise unary/binary fills/copy were added previously and already exist.

javedabsar1 committed:
[MLIR][Linalg] Add pass to convert linalg.generic back to named ops

Existing `-linalg-generalize-named-ops` lowers named ops to
linalg.generic. This patch adds `--linalg-specialize-generic-ops`
which converts, where possible, linalg.generic back to named ops.
Also, it adds patterns to recognize contractions which can be
specialized from linalg.generic to named op:
 `linalg.{batch_}?matmul{_transpose_(a|b)}?`
Patterns to recognize elementwise unary/binary fills/copy
were added previously and already exist.

adam-smnk reviewed (commented):
Thanks for taking a stab at this utility. It will definitely come in handy.
Quick initial pass for now.

javedabsar1 committed:
[MLIR][Linalg] Address review comments

javedabsar1 commented:
> Thanks for taking a stab at this utility. It will definitely come in handy. Quick initial pass for now.

Thanks.

javedabsar1 commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -107,7 +107,7 @@ isaElemwiseSingleUnaryOrBinaryOpInterface(linalg::GenericOp genericOp,
                                           unsigned arity) {
   // Check all loops are parallel, and have only tensor semantics.
```
Fixed . thanks.

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,175 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  //      (i.e. it is a permutation on RES and RHS and does not appear in LHS).
+  //   3. The k dimension appears as a permutation on LHS and RHS.
+  //   4. m, n and k appear only once in any given indexing.
+  //   5. Optional batch dimensions that appear in all operands are captured.
```
Thanks for pointing out. I will try doing that and few others - matvec mmt4d vecmat - in next diff.

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,175 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+
+  auto batchSize = dims.batch.size();
+  if (indexingMaps[0].getNumDims() != batchSize + 3) {
+  }
```
Fixed.

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,175 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+    return replaceWithMatmulVariant<MatmulTransposeAOp>(rewriter, genericOp);
+  if (b == IndexMatchResult::Transposed)
+    return replaceWithMatmulVariant<MatmulTransposeBOp>(rewriter, genericOp);
+  return replaceWithMatmulVariant<MatmulOp>(rewriter, genericOp);
```
Havent attempt it yet. Will do so next patch.

javedabsar1 commented on mlir/test/Dialect/Linalg/roundtrip-linalg-named-ops.mlir:
```diff
@@ -0,0 +1,49 @@
+// RUN: mlir-opt %s -linalg-generalize-named-ops | mlir-opt --linalg-specialize-generic-ops | FileCheck %s
+
+func.func @roundtrip_matmul(%arg0: tensor<?x?xf32>, %arg1: tensor<?x?xf32>, %arg2: tensor<?x?xf32>) -> tensor<?x?xf32> {
```
yes indeed. Fixed. thanks.

javedabsar1 commented on mlir/test/Dialect/Linalg/roundtrip-linalg-named-ops.mlir:
I just chose randomly from each - unary/binary/mix.

banach-space replied:
Let's avoid "random" :) In particular, how are we meant to know that this works for all Ops if we are only testing a random subset? if you feel that a sub-set of Linalg Ops would be sufficient, then please document through comments and/or test function names.

Also, would you mind splitting "unary",  "binary" and "mix" cases with big bold comments? Provided that that's how you want to categorise these tests.

Btw, same comment for the other test file.

javedabsar1 replied:
"please document through comments and/or test function names" - Done.
"unary", "binary" and "mix" - Done.

javedabsar1 commented:
Thanks for reviewing

MaheshRavishankar commented on mlir/test/Dialect/Linalg/roundtrip-linalg-named-ops.mlir:
```diff
@@ -0,0 +1,49 @@
+// RUN: mlir-opt %s -linalg-generalize-named-ops | mlir-opt --linalg-specialize-generic-ops | FileCheck %s
```
It is probably better to test the specialization by itself to avoid unnecessary coupling.

javedabsar1 replied:
Have added tests to test separately.
-  mlir/test/Dialect/Linalg/transform-op-specialize_matmul.mlir 
- mlir/test/Dialect/Linalg/specialize-generic-ops.mlir

I added 'round-trip' test based on my own previous 'learning' experience (e.g. in LLVM/MLIR) where beginner looks at tests to understand/experiment. The round-trip brings home the feature/tools/options and the concept:  named => generic => named.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Linalg/Passes.td:
```diff
@@ -104,6 +104,11 @@ def LinalgGeneralizeNamedOpsPass : Pass<"linalg-generalize-named-ops"> {
   let dependentDialects = ["linalg::LinalgDialect"];
 }
 
+def LinalgSpecializeGenericOpsPass : Pass<"linalg-specialize-generic-ops"> {
```
If this is already the case in this file, please keep things alphabetically sorted.

javedabsar1 replied:
It doesnt seem sorted. So kept it specialize nextto generalize.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -100,5 +280,31 @@ FailureOr<LinalgOp> mlir::linalg::specializeGenericOp(RewriterBase &rewriter,
+void LinalgSpecializeGenericOpsPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  populateLinalgGenericOpsSpecializationPatterns(patterns);
+  (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
```
Please dont ignore the result here. Catch the error and raise failure.

javedabsar1 replied:
Ping!
made changes based on comments.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,176 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+// This check skips the `offset` batch indices and focuses on the matmul part.
+static IndexMatchResult matchOperandMap(AffineMap m, unsigned offset,
+                                        unsigned i, unsigned j) {
+  auto expr_ei = dyn_cast<AffineDimExpr>(m.getResults()[offset]);
+  auto expr_ej = dyn_cast<AffineDimExpr>(m.getResults()[offset + 1]);
```
camelBack: https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide

banach-space commented on mlir/test/Dialect/Linalg/roundtrip-linalg-named-ops.mlir:
```diff
@@ -0,0 +1,49 @@
+}
+
+// CHECK-LABEL: @matmul
+// CHECK-SAME: %[[A:.+]]: tensor<?x?xf32>, %[[B:.+]]: tensor<?x?xf32>, %[[Out:.+]]: tensor<?x?xf32>) -> tensor<?x?xf32>
```
Big "+1" to using "A" and "B" for inputs (you could also use "C" for the output). Could you use similar names in the actual function names? For the sake of consistency and readability. Otherwise it's a bit odd that in one place you use `%arg0` and elsewhere `%[[A:.+]]`.

banach-space commented on mlir/test/Dialect/Linalg/roundtrip-linalg-named-ops.mlir:
```diff
@@ -0,0 +1,49 @@
+
+// -----
+
+func.func @exp(%arg: memref<7x14x21xf32>, %out: memref<7x14x21xf32>) {
```
This is the first test to use static shapes. Is the split into static/dynamic shapes significant here? if not, I suggest duplicating the first test so that there's static and dynamic variant (as a sanity check). Following from that, just stick with dynamic shapes.

qedawkins commented on mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:
```diff
@@ -1395,6 +1395,24 @@ struct LinalgGeneralizationPattern
 
+struct LinalgSpecializationPattern
+    : public OpInterfaceRewritePattern<LinalgOp> {
+  using OpInterfaceRewritePattern<LinalgOp>::OpInterfaceRewritePattern;
```
This can just be an `OpRewritePattern<GenericOp>` instead of doing a cast from LinalgOp.

javedabsar1 replied:
Awesome suggestion. simplifies. Fixed.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:
```diff
@@ -1546,6 +1564,11 @@ void populateLinalgTilingCanonicalizationPatterns(RewritePatternSet &patterns);
 /// linalg.generic ops.
 void populateLinalgNamedOpsGeneralizationPatterns(RewritePatternSet &patterns);
 
+/// Populates `patterns` with patterns to convert linalg.generic ops to named
+/// ops where possible.
+void populateLinalgGenericOpsSpecializationPatterns(
```
Added explanation in comment section.

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,176 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+// affine map per operand capture more semantics than is possible with
+// named op (which has implicit map interpreted via name).
+//
+// But a named linalg matmul variant that was 'generalized' should be
+// convertible back to named op here.
```
Removed that comment from text. We named->generic->named will.

javedabsar1 committed:
[MLIR][Linalg] More review comments changes.

javedabsar1 commented:
Gentle Ping.

rengolin commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+// not  the same as linalg.generic above.
+namespace {
+enum class IndexMatchResult {
+  Match = 0,  // identity map.
```
nit: I'd call it `Identity`

javedabsar1 replied:
Yeah. its a taste issue. the class is '*Match*'. there is another field mis-match..

banach-space replied:
[nit] +1 to "identity" :)

rengolin commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  //  <reduce> is an addition). All operands of all operations may be supplied
+  //  through a chain of side effect-free unary operations, such as casts,
+  //  which is denoted as `cu` above.
+  if (!mlir::linalg::detail::isContractionBody(
```
Does this check that the first and second ops are linked in a chain?

Ie. that at least one operand in the second is the return value of the first?

And that the return value of the second is the actual `yield`ed value?

javedabsar1 replied:
@rengolin yes. The isContraction checks that the body has 3 args;  yield is 1 value that is coming from a 'reductionOp' that has 2 inputs 1 output; the reductionLHS/RHS are block-arguments that may have had unary operations (e.g. cast) applied to them.

rengolin commented on mlir/test/Dialect/Linalg/specialize-generic-ops.mlir:
```diff
@@ -0,0 +1,37 @@
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
```
Can you add tests for matmul on the pass, too?

javedabsar1 replied:
Done.

rengolin commented on mlir/test/Dialect/Linalg/specialize-generic-ops.mlir:
```diff
@@ -0,0 +1,37 @@
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
```
Can you add tests for matmul on the pass, too?

javedabsar1 replied:
Done.

rengolin commented on mlir/test/Dialect/Linalg/transform-op-specialize_matmul.mlir:
```diff
@@ -0,0 +1,148 @@
+#map = affine_map<(d0, d1, d2) -> (d0, d2)>
+#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
+#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
+func.func @specialize_matmul(%arg0: tensor<?x?xf32>, %arg1: tensor<?x?xf32>, %arg2: tensor<?x?xf32>) -> tensor<?x?xf32> {
```
Can you add element wise for transforms, too?

javedabsar1 replied:
It was added separately when i did elemwise specialization : look for  transform-op-specialize_elemwise_*.mlir

rengolin reviewed (commented):
Are you planning to add support for broadcast/reduction/type cast soon?

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+          }))
+    return failure();
+
+  // Finds 2 parallel (m and n) and 1 reduction (k) dimension candidates that
+  // form a matmul subcomputation. These dimensions are such that:
+  //   1. The m dimension is involved in an outer-product along LHS
+  //      (i.e. it is a permutation on RES and LHS and does not appear in RHS).
+  //   2. The n dimension is involved in an outer-product along RHS
+  //      (i.e. it is a permutation on RES and RHS and does not appear in LHS).
+  //   3. The k dimension appears as a permutation on LHS and RHS.
+  //   4. m, n and k appear only once in any given indexing.
+  //   5. Optional batch dimensions that appear in all operands are captured.
+  auto res = inferContractionDims(genericOp);
```
No need to repeat the original comment: https://mlir.llvm.org/doxygen/namespacemlir_1_1linalg.html#aa2fe10e20900f7c49da8d51805f9e9f0

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  assert(succeeded(res) && "unexpected failure to infer contraction dims");
+  auto dims = *res;
+
+  // Other than `batch`, other dim sizes must be 1 for linalg.*_matmul_*.
```
Why? Surely in practice M,N and K will be =! 1?

javedabsar1 replied:
contraction can be along more than one dim as far as linalg contraction is concerned.

banach-space replied:
Now I understand where the confusion is coming from. For me, "dim sizes" in `A = M x K` are `M` and `K`. Whereas this is checking e.g. the number of "contraction dims corresponding to K as inferred by `inferControctionDims`"? So, `M` and `K` can indeed be "1". Could you clarify in the comment?

Also, are you able to add a negative test to exercise this check?

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  Mismatch    // none of the above.
+};
+
+// Consider the A matrix in `C[M,N] = A[M,K] * B[K,N]`. Below, we
+// check whether the index map of A is identity (match), transposed, or
+// something completely different (mis-match).
+// The naming and explanation is in terms of A, but the function checks
+// effectively maps for all A, B, C i.e. <M,N>, <M, K>, <K,N>.
```
[nit] Any chance to drop references to A, B and C? This seems like a very generic utility, but the naming seems to over index on a very specific use-case. For example, why not use `expectedPosDim1` and `expectedPosDim2`? That would be clearer to me.

javedabsar1 replied:
OK i tried following your suggestion and then it reads more confusing because I have to then s/exprOfM/exprPosDim1.
The idea was one reads it with one thing in mind, e.g. A in  A[m,k] * B[k,n] . Then once one gets it, person gets the hang of it and generalized. Instead of getting confused over long names.
I added more comments to make it easier read.

banach-space replied:
I understand what you are trying to achieve here, but I still find the documentation confusing 😅 I appreciate that that's a subjective matter, but IMHO a specific example should be use to complement a generic description rather than replace one.

Here's a complete suggestion. Note, I replaced `Dim1` and Dim2` from my original suggestion with `RowDim` and `ColDim` (IIUC, that's what these are):
```cpp
// Checks whether the input Affine `map` contains two consecutive dims that can
// be interpreted as accessing a 2D matrix. It is assumed that the row and
// column dimension are located next to each other (in this order) and start at
// `rowDimIdx` in the input map.
//
// YOUR SPECIFIC EXAMPLE WITH MATRIX A <<HERE>>
static IndexMatchResult matchOperandMap(AffineMap map, unsigned rowDimIdx,
                                        unsigned expectedPosOfRowDim,
                                        unsigned expectedPosOfColDim) {
  // Get the matrix multiply indices. They are past the batch indices.
  auto exprOfRowDim = map.getResults()[rowDimIdx];
  auto exprOfColDim = map.getResults()[rowDimIdx + 1];

  // They should be pure dim ids.
  if (exprOfRowDim.getKind() != AffineExprKind::DimId ||
      exprOfColDim.getKind() != AffineExprKind::DimId)
    return IndexMatchResult::Mismatch;

  auto posRowDim = cast<AffineDimExpr>(exprOfRowDim).getPosition();
  auto posColDim = cast<AffineDimExpr>(exprOfColDim).getPosition();

  if (expectedPosOfRowDim == posRowDim && expectedPosOfColDim == posColDim)
    return IndexMatchResult::Match;

  if (expectedPosOfRowDim == posColDim && expectedPosOfColDim == posRowDim)
    return IndexMatchResult::Transposed;

  return IndexMatchResult::Mismatch;
}
```

Feel free to re-use (and/or change). I just feel that we shouldn't be referring to `batchSize` and/or "dimension M"/"dimension K" in such a generic hook. For example, from the point of view of this hook it doesn't matter what the batch size is, neither does what "M" and "K" are.

HTH

javedabsar1 replied:
Thanks. Will amend based on your suggestion :)

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  return IndexMatchResult::Mismatch;
+}
+
+//  All the variants `linalg.{batch_}?matmul{_transpose_a | _transpose_b}?`
+//  have same number of input/output.
```
done

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,195 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+  //  contraction (in the canonical case, <elemwise> is a multiplication and
+  //  <reduce> is an addition). All operands of all operations may be supplied
+  //  through a chain of side effect-free unary operations, such as casts,
+  //  which is denoted as `cu` above.
```
deleted. I thought it was important to emphasis in this context. but yeah. people can read the LinalgInterfaces directly.

javedabsar1 committed:
[MLIR][Linalg] More changs based on more comments

rengolin commented:
Thanks @javedabsar1, I have exhausted my questions and will let @banach-space finish the review and approve.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
````diff
@@ -58,6 +68,181 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+// It is possible that a linalg.generic may be implementing a matmul but not
+// in a straight-forward way e.g. below is matrix multiply over some slice
+// ```
+//  %0 = linalg.generic {
+//          indexing_maps = [affine_map<(d0, d1, d2) -> (3, d1, d0)>,
+//                           affine_map<(d0, d1, d2) -> (d0, 5, d2)>,
+//                           affine_map<(d0, d1, d2) -> (d2, d1, 13)>],
+//          iterator_types = ["parallel", "parallel", "parallel"]}
+//          ins(%A, %B : tensor<20x20x20xf32>,  tensor<20x20x20xf32>)
+//          outs(%C : tensor<20x20x20xf32>) {
+//             ^bb0(%a: f32, %b: f32, %c : f32):
+//                %mul = arith.mulf %a, %b : f32
+//                %add = arith.addf %mul, %c : f32
+//                linalg.yield %add : f32
+//       } -> tensor<20x20x20xf32>
````
Could you use it to write a negative test?

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -58,6 +68,181 @@ static bool areBinOpsSwapped(GenericOp genericOp) {
+      }))
+    return failure();
+
+  auto batchSize = dims.batch.size();
```
To me, "batch size" would be `2` in this example:
```mlir
 linalg.batch_matmul ins(%[[A]], %[[B]] : tensor<2x16x8xf32>, tensor<2x8x16xf32>) outs(%[[Out]] : tensor<2x16x16xf32>) -> tensor<2x16x16xf32>
```

Whereas `batchSize = dims.batch.size()` is "the number of batch dims", so `1`. I suggest renaming `batchSize` as `numOfBatchDims`.

Also, would the number of batch dims be ever != 1?

banach-space reviewed (commented):
I've finally managed to scan the whole thing. Overall LG, thanks! And thanks again for the contribution!

One thing that's still missing - some negative tests. I've made a few specific suggestions inline. You could also add one for e.g. `linalg.matvec` and mark it as TODO :) 

I've also left a couple of more small suggestion inline, but nothing major.

Also, looks like most files in "mlir/test/Dialect/Linalg" use hyphen (`-`) rather than underscore in filenames (`_`). For consistency:
* "transform-op-specialize_matmul.mlir" -> "transform-op-specialize-matmul.mlir"

Same suggestion for "transform-op-specialize_elemwise_binary.mlir" and "transform-op-specialize_elemwise_unary.mlir" (the only other 2 files with inconsistent naming)

javedabsar1 committed:
[MLIR][Linalg] changes based on review comments
- renaming batchSize as numOfBatchDims
- example of multiple contraction dims corresponding to K as inferred by inferControctionDims
- add linalg.matvec and mark it as TODO
- use hyphen (-) rather than underscore in filenames (_).
- implement banach suggestion to replace M K in explanation with expectedPosOfRowDim etc

javedabsar1 commented:
Made all changes requested as much as possible.

javedabsar1 committed:
[MLIR][Linalg] Fix clang format complain

banach-space reviewed (approved):
LGTM, thank you for addressing my comments and for seeing this through - super handy!

I've left one nit - feel free to ignore. I believe that you've already addressed comments from other reviewers, so this can be merged. Please don't forget to fix the formatting ;-)

Thanks again!

11happy commented:
@javedabsar1  Does this pass change linalg generic which is just copying tensor from input to output, to linalg.copy?
![image](https://github.com/user-attachments/assets/29982cb2-179e-461f-a050-02b3972078e8)
